src/C++/SSLSocketConnection.cpp

```diff
@@ -359,7 +372,14 @@ EXCEPT ( SocketRecvFailed )
       {
         errno = EINTR;
         size = 0;
-
+        
+        if (errCodeSSL == SSL_ERROR_WANT_WRITE)
+        {
+            Locker locker(m_mutex);
+            m_readFromSocketNeedsToWriteData = true;
+            subscribeToSocketWriteAvailableEvents();
+        }
```

Comment:
Here, readFromSocket's SSL_read failed because the OpenSSL statemachine wants to write data (eg for key renegotiation). To help with this problem, we set the m_readFromSocketNeedsToWriteData flag and signal the monitor to notify us with onWrite events. onWrite will then check this flag and call this method so the SSL_read can resume.